(Towards #714) Initial implementation of ChecksumTrans with test

LonelyCat124 opened the pull request:
Initial implementation of a transformation that adds the ability to automatically add checksums around a region of code.

LonelyCat124 commented:
@sergisiso @arporter ready for a first look.

hiker commented:
Could you provide some details about this transformations? To me it sounds like it is doing pretty much what the `ReadOnlyVerifyTrans` is doing (except for a different set of variables, but that is pretty trivial to change). This would on the other hand require to add a psydata library at compile time (though it then also supports 64 bit checksums for int, real, char, logical, ...)

LonelyCat124 commented:
Hi @hiker , we weren't sure if the other psydata API  implementation supported non-lfric code (as it references invoke and kernels) so we threw this together super quickly. Happy to use the other if it would work though

hiker commented:
You can always just check the library directory:
```
PSyclone/lib/read_only$ ls -l
total 348
...
drwxr-xr-x 2 joerg joerg   4096 Mar 13 19:57 dl_esm_inf
drwxr-xr-x 2 joerg joerg   4096 Mar 13 19:57 generic
drwxr-xr-x 2 joerg joerg   4096 Mar 13 19:57 lfric
```
:)

I can't say if this is actually more work - you would have to create the psydata transformation, and likely create a different implementation of the library (since the old library compares checksums, you want to print them??).
It also means that there is an additional dependency at compile time. So `sum()` might be easier - then again, you are doing floating point sums, while the psydata lib computes 64-bit integer checksums. I am not certain what you need this for :)

LonelyCat124 commented:
Known issues highlighted by @JonElsey, he's got some initial fix I think:  

- [ ]  Character/Logical arrays (and maybe UnknownFortranType generally) should be ignored.
- [ ] Allow specification of "print" as an option, e.g. umPrint for UKCA.

LonelyCat124 commented:
@sergisiso @arporter I think I resolved the issue - I'm not going to worry about Umprint (or other things) for now since that won't be part of LFRic as I understand it.
I forbid any symbol that dosen't have a INTEGER or REAL datatype intrinsic from being checksummed (and check that the symbol is a DataSymbol before i check the datatype).

LonelyCat124 commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,166 @@
+from psyclone.core import VariablesAccessInfo
+from psyclone.psyir.nodes import Assignment, Node, Reference, Routine
+from psyclone.psyir.transformations.region_trans import RegionTrans
+from psyclone.psyir.frontend.fortran import FortranReader
```
Fixed

LonelyCat124 replied:
Oops, fixed there and in the docs.

LonelyCat124 commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,166 @@
+
+class DebugChecksumTrans(RegionTrans):
+    '''
+    Creates a set of checksums (written via print) for all written to arrays
```
Fixed.

LonelyCat124 commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,166 @@
+        enddo
+      enddo
+      PSYCLONE_INTERNAL_line_ = __LINE__
+      PRINT *, "checksums from mysubroutine at line:", PSYCLONE_INTERNAL_line_\
```
Yeah, added.

LonelyCat124 commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,166 @@
+
+    def apply(self, node: Union[Node, List[Node]], options=None) -> None:
+        '''
+            Applies the checksum transformation to the provided node(s).
```
Fixed

LonelyCat124 commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,166 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2017-2025, Science and Technology Facilities Council.
```
Fixed

LonelyCat124 commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,166 @@
+        for node in checksum_nodes:
+            parent.addchild(node, position+1)
+
+        internal_line = \
```
Added

LonelyCat124 commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,166 @@
+            parent.addchild(node, position+1)
+
+        internal_line = \
+            node_list[0].ancestor(Routine).symbol_table.find_or_create(
```
Yeah, done

LonelyCat124 commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,166 @@
+                "PSYCLONE_INTERNAL_line_", symbol_type=DataSymbol,
+                datatype=INTEGER_TYPE,
+                )
+        line = node_list[0].ancestor(Routine).symbol_table.find_or_create(
```
Have one comment now explaining this section and why.

LonelyCat124 commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,166 @@
+                f', PSYCLONE_INTERNAL_line_ + 1',
+                node_list[0].ancestor(Routine).symbol_table
+                )
+        # Remove the comment about this being a code block.
```
Sure, seems like a good idea.

LonelyCat124 commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,166 @@
+            :type options: Optional[Dict[str, Any]]
+
+        '''
+        self.validate(node)
```
Yes, but I can just disable `node-type-check` for this transformation I think, since this transformation takes no options..

LonelyCat124 commented on src/psyclone/tests/psyir/transformations/debug_checksum_trans_test.py:
```diff
@@ -0,0 +1,116 @@
+# -----------------------------------------------------------------------------
+# BSD 3-Clause License
+#
+# Copyright (c) 2021-2025, Science and Technology Facilities Council.
```
Fixed

LonelyCat124 commented on src/psyclone/tests/psyir/transformations/debug_checksum_trans_test.py:
```diff
@@ -0,0 +1,116 @@
+from psyclone.psyir.transformations import DebugChecksumTrans
+
+
+def test_checksum(fortran_reader, fortran_writer):
```
Added

LonelyCat124 commented on src/psyclone/tests/psyir/transformations/debug_checksum_trans_test.py:
```diff
@@ -0,0 +1,116 @@
+
+    DebugChecksumTrans().apply(psyir.children[0].children[0])
+
+    correct = """subroutine test()
```
Done

LonelyCat124 commented on src/psyclone/tests/psyir/transformations/debug_checksum_trans_test.py:
```diff
@@ -0,0 +1,116 @@
+
+  a(:) = 1
+  b(:) = 2
+  PSYCLONE_INTERNAL_line_ = __LINE__
```
Done

LonelyCat124 commented on src/psyclone/tests/psyir/transformations/debug_checksum_trans_test.py:
```diff
@@ -0,0 +1,116 @@
+end subroutine test
+"""
+    out = fortran_writer(psyir)
+    assert correct == out
```
Added two more tests to cover these - unresolved types has its own test, the rest are in one.

LonelyCat124 commented:
@arporter This has undergone a bit of a rewrite - the output is similar to before but it now uses .datatype and looks at assignments in the specified section.

arporter commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,203 @@
+        # Loop through the assignments and find the arrays
+        for assign in assigns:
+            if isinstance(assign.lhs, StructureAccessorMixin):
+                # Find the last member.
```
You shouldn't have to do this - `assign.lhs.datatype` *should* be all that you need? If it isn't, please could you explain what the problem is?

LonelyCat124 replied:
Added comments to why this is done.

arporter replied:
Great, thanks.

arporter commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,203 @@
+        checksum_nodes = []
+        freader = FortranReader()
+        for lhs in writes:
+            # Need to convert the lhs to a full range variant.
```
The existing data-extraction code must also do this so may have functionality you can re-use. I think I also had to do it for inlining but that's probably buried in the InlineTrans. In fact, I've had a look and it is very similar but probably not similar enough to be useful.

LonelyCat124 replied:
I think the code that does this here is applying to to specific sections and the code that actual does the replacing is very minimal, i.e. just:
```
            # Need to convert the lhs to a full range variant.
            for i in range(len(array_bit.indices)):
                new_index = array_bit.get_full_range(i)
                array_bit.indices[i].replace_with(new_index)
```

I extracted that part out to work on this array_bit instead of being repeated.

arporter commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,203 @@
+                while isinstance(member, StructureAccessorMixin):
+                    member = member.member
+                datatype = assign.lhs.datatype
+                for i in range(len(member.indices)):
```
Unfortunately, I think this is assuming that you always have, e.g. `myvar%something%else(i,j) = ...` while you could well have `myvar%something(i,j)%else = ...`. Rather than get bogged down in this, I suggest we just exclude structure references for now. Alternatively, you could look at using VariablesAccessInfo?

arporter replied:
Actually, rather than exclude all types, we could just exclude those that this code currently doesn't work for (e.g. `myvar%something(i,j)%else`).

arporter commented on src/psyclone/tests/psyir/transformations/debug_checksum_trans_test.py:
```diff
@@ -0,0 +1,154 @@
+
+    correct = "integer :: PSYCLONE_INTERNAL_line_"
+    assert correct in out
+    correct = """PSYCLONE_INTERNAL_line_ = __LINE__
```
Could you extend this slightly to either include the `b(:) = ...` or the `end subroutine` so that we know the statements are in the right location.

LonelyCat124 replied:
Done. Also added a test where we didn't do the entire schedule.

arporter reviewed (changes requested):
Thanks Aidan. I'd forgotten just how much 'fun' derived types were - you'll see there are some issues with what you've done. Hopefully we can just exclude more complex cases and that will be good enough for most of the time. I was hoping we had some general functionality to help but I don't think we do. Possibly it's something we should add to e.g. `StructureAccessorMixin` in the future.

In addition to this, please could you update the Transformations section of the UserGuide and also check that the Reference Guide includes the output for this new transformation.

LonelyCat124 commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,203 @@
+          a(i,j) = a(i,j) + i - j
+        enddo
+      enddo
+      PSYCLONE_INTERNAL_line_ = __LINE__
```
Added a warning about this (and checked the docs that is appears correctly).

I'm not sure where the current output file name could go other than the Config, though I'm not sure if that is quite the right place - also the output file name can be None still and this transformation should probably be ok with that (e.g. for testing/looking at outputs etc.).

LonelyCat124 replied:
I also put the warning in the User guide as well as for the developers guide in the docstring.

arporter replied:
Thanks Aidan. I think that will do for now.

LonelyCat124 commented:
@arporter Sending this back for review - some of the changes are "unaddressed", but I added comments as to why the code is as it is and we discussed in teams.

> In addition to this, please could you update the Transformations section of the UserGuide and also check that the Reference Guide includes the output for this new transformation.
I added the relevant section to the UserGuide (with a warning) and checked the ReferenceGuide (as far as I understand) shows the output expected.

arporter commented on doc/user_guide/transformations.rst:
```diff
@@ -316,6 +316,16 @@ can be found in the API-specific sections).
+      :members: apply
+      :no-index:
+
+.. warning::  This transformation generates code that will only work with .F90
```
You don't need this one as the documentation generated from the docstring by `.. autoclass` includes the one from there.

LonelyCat124 replied:
Ah ok, I'll remove it.

arporter commented on src/psyclone/psyir/transformations/debug_checksum_trans.py:
```diff
@@ -0,0 +1,227 @@
+                    # which arrays or indexes to sum over.
+                    if (isinstance(member, ArrayMixin) and
+                            member.member is not None):
+                        multiple_arrays = True
```
Codecov complains about L146. You could possibly simplify this by using `get_signature_and_indices()` and examining the list of indices - every entry except the last should be empty if we are to support creating a checksum.

arporter replied:
OK. That's not much (any?) simpler and you've got coverage so I'm happy :-)

arporter reviewed (changes requested):
Almost there now. Just a little bit of tidying and then this can go on.

arporter reviewed (approved):
All good now, thanks very much Aidan.
I've checked that the updated docs build OK.
Ideally we'd have an example to show this transformation and/or add it to the tutorial material but that can be future work.
